File: 0227. Basic Calculator II/main.cpp
#include <iostream>
#include <stack>
using namespace std;

/**
 * I: s = "3+2*2"
 * O: 7
 *
 * I: s = " 3/2 "
 * O: 1
 *
 * I: s = " 3+5 / 2 "
 * O: 5
 *
 * I: s = "22 - 3 * 5"
 * O: 7
 */

int calculate(string s);

int main(int argc, char const *argv[]) {
    string s = "3/2";
    cout << calculate(s) << endl;
    return 0;
}

int calculate(string s) {
    int len = s.length();
    if (len == 0) return 0;
    stack<int> stack;
    int currentNumber = 0;
    char operation = '+';
    for (int i = 0; i < len; i++) {
        char currentChar = s[i];
        if (isdigit(currentChar)) {
            currentNumber = (currentNumber * 10) + (currentChar - '0');
        }
        if (!isdigit(currentChar) && !iswspace(currentChar) || i == len - 1) {
            if (operation == '-') {
                stack.push(-currentNumber);
            } else if (operation == '+') {
                stack.push(currentNumber);
            } else if (operation == '*') {
                int stackTop = stack.top();
                stack.pop();
                stack.push(stackTop * currentNumber);
            } else if (operation == '/') {
                int stackTop = stack.top();
                stack.pop();
                stack.push(stackTop / currentNumber);
            }
            operation = currentChar;
            currentNumber = 0;
        }
    }
    int result = 0;
    while (!stack.empty()) {
        result += stack.top();
        stack.pop();
    }
    return result;
}

File: 0227. Basic Calculator II/main.js
/**
 * @param {string} s
 * @return {number}
 */
let calculate = function (s) {
  let len = s.length;
  if (len === 0) return 0;
  let stack = [];
  let operator = "+";
  let currentNumber = 0;
  for (let i = 0; i < len; i++) {
    let curChar = s[i];
    // console.log(curChar);
    if (curChar >= "0" && curChar <= "9") {
      currentNumber = currentNumber * 10 + Number(curChar);
    }
    if (
      (!(curChar >= "0" && curChar <= "9") && curChar !== " ") ||
      i === len - 1
    ) {
      let top;
      switch (operator) {
        case "+":
          stack.push(currentNumber);
          //   console.log(stack);
          break;
        case "-":
          stack.push(-currentNumber);
          //   console.log(stack);
          break;
        case "*":
          top = stack.pop();
          stack.push(top * currentNumber);
          //   console.log(stack);
          break;
        case "/":
          top = stack.pop();
          stack.push(parseInt(top / currentNumber));
          //   console.log(stack);
          break;

        default:
          break;
      }
      currentNumber = 0;
      operator = curChar;
    }
  }
  console.log(stack);
  return stack.reduce((a, b) => {
    return a + b;
  });
};

const s = "3/2";
console.log(calculate(s));
